packages/block-editor/src/components/block-list/block-wrapper.js

```diff
@@ -184,43 +192,69 @@ const BlockComponent = forwardRef(
 			if ( ( buttons || which ) === 1 ) {
 				onSelectionStart( clientId );
 			}
+		}
+
+		if ( isSelected ) {
+			// Only allow shortcuts when a blocks is selected and not locked.
+			if ( ! isLocked ) {
+				ref.current.addEventListener( 'keydown', onKeyDown );
+			}
+
+			// Only allow selection to be started from a selected block.
+			ref.current.addEventListener( 'mouseleave', onMouseLeave );
+		}
+
+		return () => {
+			if ( isSelected ) {
+				if ( ! isLocked ) {
+					ref.current.removeEventListener( 'keydown', onKeyDown );
+				}
+
+				ref.current.removeEventListener( 'mouseleave', onMouseLeave );
+			}
 		};
+	}, [
+		isSelected,
+		isLocked,
+		insertDefaultBlock,
+		removeBlock,
+		onSelectionStart,
+	] );
 
-		const htmlSuffix =
-			mode === 'html' && ! __unstableIsHtml ? '-visual' : '';
-		const blockElementId = `block-${ clientId }${ htmlSuffix }`;
-
-		return (
-			<TagName
-				// Overrideable props.
-				aria-label={ blockLabel }
-				role="group"
-				{ ...omit( wrapperProps, [ 'data-align' ] ) }
-				{ ...props }
-				id={ blockElementId }
-				ref={ wrapper }
-				className={ classnames(
-					className,
-					props.className,
-					wrapperProps && wrapperProps.className
-				) }
-				data-block={ clientId }
-				data-type={ name }
-				data-title={ blockTitle }
-				// Only allow shortcuts when a blocks is selected and not locked.
-				onKeyDown={ isSelected && ! isLocked ? onKeyDown : undefined }
-				// Only allow selection to be started from a selected block.
-				onMouseLeave={ isSelected ? onMouseLeave : undefined }
-				tabIndex="0"
-				style={ {
-					...( wrapperProps ? wrapperProps.style : {} ),
-					...( props.style || {} ),
-				} }
-			>
-				{ children }
-			</TagName>
-		);
-	}
+	const htmlSuffix = mode === 'html' && ! __unstableIsHtml ? '-visual' : '';
+
+	return {
+		...wrapperProps,
+		...props,
+		ref,
+		id: `block-${ clientId }${ htmlSuffix }`,
+		tabIndex: 0,
+		role: 'group',
+		'aria-label': blockLabel,
+		'data-block': clientId,
+		'data-type': name,
+		'data-title': blockTitle,
+		className: classnames(
+			'wp-block',
+			className,
+			props.className,
+			wrapperProps.className
+		),
+		style: { ...wrapperProps.style, ...props.style },
+	};
+}
+
+const BlockComponent = forwardRef(
```

Comment:
This is cool, when do you think we should remove this? should we add a deprecated call with a version?

Reply:
Since it's experimental, maybe we can just remove it. Depending on what we do, we might have to leave it since some block components are still classes. Maybe the ref idea could replace it though.

Reply:
We could yes, I saw some third-party usage already. So let's be nice to them and deprecate for a release or two.